### PR TITLE
SPHINCS+ optimized implementations

### LINKED ISSUE #249: Improve documentation for hashing API in header file sha2.h
_Originally posted by @thomwiggers in https://github.com/PQClean/PQClean/pull/247#issuecomment-545841257_

### PR DESCRIPTION
**Pull request by thomwiggers**
Include SPHINCS+ optimized implementations: AVX2 and AESNI.

Also updates the `clean` versions, because they're being generated from a shared codebase.

Supersedes #247: closes #247 when merged. Also fixes #249 

#### Manually checked properties

* [x] No stringification macros
* [x] Output-parameter pointers in functions are on the left
* [x] Negative return values on failure of API functions (within restrictions of FO transform).
* [x] `const` arguments are labeled as `const`
* [x] variable declarations at the beginning (except in `for (size_t i=...`)
* Optional:
  * [ ] All integer types are of fixed size, using `stdint.h` types (including `uint8_t` instead of `unsigned char`)
  * [ ] Integers used for indexing are of size `size_t`

### CODE REVIEW

**Comment by thomwiggers** [comment]
Shit, this should've been a draft PR.

**Comment by thomwiggers** [comment]
* [ ] Tested on Windows
  * [ ] 64-bit
  * [ ] 32-bit
* [ ] Tested on Linux:
  * [x] amd64
    * [ ] Clang
    * [ ] GCC
  * [ ] i386
    * [ ] Clang
    * [ ] GCC
  * [ ] arm32
    * [ ] Clang
    * [ ] GCC
  * [ ] arm64
    * [ ] Clang
    * [ ] GCC
  * [ ] ppc
    * [ ] Clang
    * [ ] GCC

**Comment by thomwiggers** [comment]
See also https://github.com/thomwiggers/sphincsplus/tree/pqcleanup

File: crypto_sign/sphincs-haraka-128f-robust/clean/address.h [hunk 1]
```diff
@@ -1,13 +1,13 @@
...
+#define PQCLEAN_SPHINCSHARAKA128FROBUST_CLEAN_ADDRESS_H
 
 #include <stdint.h>
 
-#define SPX_ADDR_TYPE_WOTS 0
-#define SPX_ADDR_TYPE_WOTSPK 1
-#define SPX_ADDR_TYPE_HASHTREE 2
-#define SPX_ADDR_TYPE_FORSTREE 3
-#define SPX_ADDR_TYPE_FORSPK 4
+#define PQCLEAN_SPHINCSHARAKA128FROBUST_CLEAN_ADDR_TYPE_WOTS 0
```
Comment (mkannwischer) [inline crypto_sign/sphincs-haraka-128f-robust/clean/address.h]: Why namespace these constants?

Comment (thomwiggers) [reply crypto_sign/sphincs-haraka-128f-robust/clean/address.h]: They always had an SPX_ prefix; which was also the prefix I chose to add to all symbols. They simply got namespaced with all the rest.

Comment (mkannwischer) [reply crypto_sign/sphincs-haraka-128f-robust/clean/address.h]: ok, I don't think this is necessary or consistently done in the other implementations in PQClean, but it doesn't hurt.

**Pull request merged**